Analyze hotspots

shubham7109 opened the pull request:
## Description
<!--
PR to add a new Kotlin sample _"SAMPLE_NAME"_ in `SAMPLE_CATEGORY` category.
-->
## Links and Data
<!--
Sample Epic: `runtime/kotlin/issues/ISSUE_NUMBER`
-->
## What To Review
<!--
-  Review the code to make sure it is easy to follow like other samples on Android
- `README.md` and `README.metadata.json` files
-->

## How to Test
<!--
Run the sample on the sample viewer or the repo.
-->

<!-- OPTIONAL
## To Discuss
-->

<!-- OPTIONAL
## Screenshots
-->

prupani-7 commented on analyze-hotspots/src/main/java/com/esri/arcgismaps/sample/analyzehotspots/components/MapViewModel.kt:
```diff
@@ -0,0 +1,171 @@
+    /**
+     * Displays an error dialog with [title] and optional [description]
+     */
+    fun showErrorDialog(title: String, description: String = "") {
```
It looks like the `showErrorDialog` fun will be used by other samples too. Should we consider moving it to common samples-lib?

shubham7109 replied:
Good point, added a `MessageDialogViewModel` to the `MessageDialog` component in the samples-lib

prupani-7 reviewed (commented):
This is a nice sample introducing new components! Good work!!

`sampleCoroutineScope` and `MessageDialog ` implementation looks neat. 
I have added some suggestions below. 

Also, once we select the "FROM" date, I think it would be good to have the "TO" date disable all dates prior to the FROM date selected, so we don't end up selecting a back date in TO DatePicker. (which would mean the date range is invalid). 

I notice, we can still `Run Analysis` with this invalid date range, a job starts but eventually errors out with below message, which isn't the error we would expect. I don't think we should try to achieve this enhancement in this PR, if its not an easy fix, but definitely worth thinking about.
<img width="333" alt="image" src="https://github.com/Esri/arcgis-maps-sdk-kotlin-samples/assets/44178002/a02adfb0-bb1d-4be2-bb03-5e402f687c8f">

prupani-7 commented:
Have you noticed this? The job loading dialog, shows progress until only 10% (never reaches 100%), jobs succeeds and adds the layer on the hotspot map.

shubham7109 commented:
@prupani-7 Thanks for the helpful feedback, I've made the respective changes here. Looking forward for your re-review :)

prupani-7 reviewed (approved):
Looks good to me!!

shubham7109 commented:
Updated this sample to also reflect the comments from #116, and modified the `MapViewState` to not be a state flow w.r.t this comment: https://github.com/Esri/arcgis-maps-sdk-kotlin-samples/pull/116#discussion_r1231896846

Thanks @prupani-7 for the review. Hopefully it's okay that I've made a few changes, so you're free to suggest any changes or could leave the approval as it is.

hud10837 commented on analyze-hotspots/src/main/java/com/esri/arcgismaps/sample/analyzehotspots/components/MapViewModel.kt:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    fun cancelGeoprocessingJob(coroutineScope: CoroutineScope) {
+        coroutineScope.launch {
```
Could you use `sampleCoroutineScope` here?

hud10837 commented on analyze-hotspots/src/main/java/com/esri/arcgismaps/sample/analyzehotspots/screens/BottomSheetScreen.kt:
```diff
@@ -0,0 +1,117 @@
+ */
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun BottomSheetScreen(
```
Since this already is only called within a bottom sheet, maybe it could be renamed to something like `DateRangeSelector`?

hud10837 commented on samples-lib/src/main/java/com/esri/arcgismaps/sample/sampleslib/components/MessageDialog.kt:
```diff
@@ -0,0 +1,110 @@
+    var dialogStatus = mutableStateOf(false)
+        private set
+
+    var messageTitle = ""
+        private set
+
+    var messageDescription = ""
```
Should these be mutable states as well?

hud10837 reviewed (changes requested):
Looks good, a few comments

shubham7109 commented on analyze-hotspots/src/main/java/com/esri/arcgismaps/sample/analyzehotspots/screens/BottomSheetScreen.kt:
```diff
@@ -0,0 +1,117 @@
+}
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Preview(showBackground = true)
```
Not really, the first `@Preview` will preview in light mode, and the second `@Preview` will preview in dark mode.

shubham7109 commented on analyze-hotspots/src/main/java/com/esri/arcgismaps/sample/analyzehotspots/screens/MainScreen.kt:
```diff
@@ -0,0 +1,105 @@
+                                )
+                            } else {
+                                sampleCoroutineScope.launch {
+                                    mapViewModel.apply {
```
The apply block is helpful here to reduce the noise from using the `mapViewModel` multiple times:
```kotlin
// create and run a geoprocessing task using date range
mapViewModel.createGeoprocessingJob(
    fromDate = mapViewModel.convertMillisToString(fromDateInMillis),
    toDate = mapViewModel.convertMillisToString(toDateInMillis),
)
```

Updated to directly just call the method 👍

hud10837 replied:
I guess it could also be an extension method on `Long`, so you could just do `fromDate = fromDateMs.toDateFormattedString()` but I'm not even sure that's better 😄

shubham7109 commented:
Thanks @hud10837 for the review. I've added the changes. Ready for your re-review!

hud10837 reviewed (approved):
LGTM 👍
